versioned/transfer/src/main/java/org/projectnessie/versioned/transfer/ExportPersist.java

```diff
@@ -106,14 +108,20 @@ private HeadsAndForkPoints scanAllReferences(Consumer<CommitObj> commitHandler)
         .queryReferences(referencesQuery())
         .forEachRemaining(
             ref -> {
-              for (Iterator<CommitObj> commitIter =
-                      commitLogic.commitLog(commitLogQuery(ref.pointer()));
-                  commitIter.hasNext(); ) {
-                CommitObj commit = commitIter.next();
-                if (!identify.handleCommit(commit)) {
-                  break;
+              Deque<ObjId> commitsToProcess = new ArrayDeque<>();
+              commitsToProcess.push(ref.pointer());
+              while (!commitsToProcess.isEmpty()) {
+                ObjId id = commitsToProcess.pop();
+                Iterator<CommitObj> commitIter = commitLogic.commitLog(commitLogQuery(id));
+                while (commitIter.hasNext()) {
+                  CommitObj commit = commitIter.next();
+                  if (identify.handleCommit(commit)) {
+                    commitHandler.accept(commit);
+                    for (ObjId objId : commit.secondaryParents()) {
+                      commitsToProcess.push(objId);
```

Comment:
Let's add a check in `IdentifyHeadsAndForkPoints` that just tests whether a commit-_ID_ has been seen (basically what `identify.handleCommit()` does at the beginning) - and only push the ID if the commit has not been seen before.
And the same check after the `.pop` above.